pkg/controller/projectreference/projectreference_controller_test.go

```diff
@@ -278,69 +280,123 @@ var _ = Describe("ProjectReference controller reconcilation", func() {
+					_, err := reconciler.Reconcile(reconcile.Request{NamespacedName: projectReferenceName})
+					Expect(err).To(HaveOccurred())
 				})
-				_, err := reconciler.Reconcile(reconcile.Request{NamespacedName: projectReferenceName})
-				Expect(err).To(HaveOccurred())
 			})
-		})
 
-		Context("When the failing to update Status to Ready", func() {
-			It("It requeues with error", func() {
-				mockKubeClient.EXPECT().Get(gomock.Any(), gomock.Any(), gomock.Any()).SetArg(2, configMap).Times(1)
-				mockGCPClient.EXPECT().ListProjects().Return([]*cloudresourcemanager.Project{{LifecycleState: "ACTIVE", ProjectId: projectReference.Spec.GCPProjectID}}, nil)
-				mockGCPClient.EXPECT().ListAPIs(gomock.Any())
-				mockGCPClient.EXPECT().EnableAPI(gomock.Any(), gomock.Any()).AnyTimes()
-				mockGCPClient.EXPECT().CreateCloudBillingAccount(gomock.Any(), gomock.Any()).Return(nil)
-				mockGCPClient.EXPECT().GetServiceAccount(gomock.Any()).Return(&iam.ServiceAccount{Email: "Some Email"}, nil).Times(2)
-				mockGCPClient.EXPECT().GetIamPolicy(gomock.Any()).Return(&cloudresourcemanager.Policy{}, nil)
-				mockGCPClient.EXPECT().SetIamPolicy(gomock.Any()).Return(&cloudresourcemanager.Policy{}, nil)
-				mockGCPClient.EXPECT().CreateServiceAccountKey(gomock.Any()).Return(&iam.ServiceAccountKey{PrivateKeyData: "dGVzdAo="}, nil)
-				mockKubeClient.EXPECT().Get(gomock.Any(), gomock.Any(), gomock.Any()).Return(errors.New("Ooops not found")).Times(1)
-				mockKubeClient.EXPECT().Create(gomock.Any(), gomock.Any()).Return(nil)
-				mockKubeClient.EXPECT().Status().Return(mockUpdater).Times(2)
-				mockUpdater.EXPECT().Update(gomock.Any(), gomock.Any()).Return(errors.New("Fake update Error")).Times(2)
-				_, err := reconciler.Reconcile(reconcile.Request{NamespacedName: projectReferenceName})
-				Expect(err).To(HaveOccurred())
+			Context("When processes the project reference correctly", func() {
+				It("It does not requeue", func() {
+					mockGCPClient.EXPECT().ListAPIs(gomock.Any()).Return(OSDRequiredAPIS, nil)
+					mockGCPClient.EXPECT().GetServiceAccount(gomock.Any()).Return(&iam.ServiceAccount{Email: "Some Email"}, nil).Times(2)
+					mockGCPClient.EXPECT().GetIamPolicy(gomock.Any()).Return(&cloudresourcemanager.Policy{}, nil)
+					mockGCPClient.EXPECT().SetIamPolicy(gomock.Any()).Return(&cloudresourcemanager.Policy{}, nil)
+					mockGCPClient.EXPECT().CreateServiceAccountKey(gomock.Any()).Return(&iam.ServiceAccountKey{PrivateKeyData: "dGVzdAo="}, nil)
+					mockKubeClient.EXPECT().Get(gomock.Any(), gomock.Any(), gomock.Any()).Return(fakeError).Times(1)
+					mockKubeClient.EXPECT().Create(gomock.Any(), gomock.Any()).Return(nil)
+					mockKubeClient.EXPECT().Status().Return(mockUpdater)
+					mockUpdater.EXPECT().Update(gomock.Any(), gomock.Any()).Return(nil)
+					_, err := reconciler.Reconcile(reconcile.Request{NamespacedName: projectReferenceName})
+					Expect(err).ToNot(HaveOccurred())
+				})
```

Comment:
This feels a bit like code duplication from below, maybe you can find a way without.
All those Its duplicate a lot of `EXPECT` calls, maybe we can (ccs as well as non-ccs) maybe they can be unified into `BeforeEach` blocks?

Reply:
Will try, thanks!

Reply:
I see there is kind of duplicate, but I don't find a good way to do it. How to put CCS and non-CCS into one `BeforeEach` block? 

We use a reference to pass `projectReference` so I must have two separate blocks for CCS and non-CCS.
``` go
mockKubeClient.EXPECT().Get(gomock.Any(), projectReferenceName, gomock.Any()).SetArg(2, *projectReference)
```